lib/puppet/http/resolver/srv.rb

```diff
@@ -1,15 +1,17 @@
 class Puppet::HTTP::Resolver::SRV < Puppet::HTTP::Resolver
-  def initialize(domain: srv_domain, dns: Resolv::DNS.new)
+  def initialize(client, domain:, dns: Resolv::DNS.new)
+    @client = client
     @srv_domain = domain
     @delegate = Puppet::Network::Resolver.new(dns)
   end
 
-  def resolve(session, name, &block)
+  def resolve(session, name, ssl_context: nil, &block)
     # Here we pass our HTTP service name as the DNS SRV service name
     # This is fine for :ca, but note that :puppet and :file are handled
     # specially in `each_srv_record`.
     @delegate.each_srv_record(@srv_domain, name) do |server, port|
-      yield session.create_service(name, server, port)
+      service = Puppet::HTTP::Service.create_service(@client, name, server, port)
+      return service if check_connection?(session, service, ssl_context: ssl_context)
     end
   end
```

Comment:
If we don't return a service by this point, we'll return an empty array. This fails when we try to log `service.url` in the debug message in the session, since `if service` is truthy when `service = []`

Reply:
Returning from inside the block is one of those sometimes-necessary-evils in ruby. If you don't add an explicit return, then the `resolve` method will return whatever `each_srv_record` returns, which following that rabbit hole, comes from `pri_hash.keys.sort` in https://github.com/puppetlabs/puppet/blob/master/lib/puppet/network/resolver.rb#L154-L156 being an empty array. So ruby sloppily returns whatever was "last" evaluated.

In order to return the first connected service, you have to add the explicit return. It's called a non-local return, since it returns from the outer `resolve` method, not the local block.

The downside is the `each_srv_record` method may not be expecting a non-local return. In this case it's ok. But if the `each_srv_record` method made a socket connection and yielded, then the non-local return would prevent the socket from being closed. To protect against that, you'll often see the method ensure the resource is closed even if the block does a non-local return or raises an exception. For example, the block form of [`Net::HTTP.start` uses an `ensure`](https://github.com/ruby/ruby/blob/cdcaf041127430b30ee4b32452762666495a732a/lib/net/http.rb#L917-L924)

The other thing with non-local returns and iterators is to make sure we sure we return an explicit value if we walk the entire list and don't find a service. So we need something like:

```ruby
def resolve(...)
  @delegate.each_srv_record(@srv_domain, name) do |server, port|
  ...
  end
  nil
end
```

That way if there are no services, we return `nil` and the session should fallback to the settings resolver.